[Dynamic buffer calculation] Fix the logic to fetch the effective speed in headroom size calculation

stephenxs opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Summary:
Fix the logic to fetch the effective speed when calculating headroom size in the test case.

Signed-off-by: Stephen Sun <stephens@nvidia.com>

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)


### Back port request
- [ ] 201911

### Approach
#### What is the motivation for this PR?
Fix the logic to fetch the effective speed when calculating headroom size in the test case.

#### How did you do it?
Originally, the effective speed is the configured speed.
However, with port auto-negotiation supported, we need to take more factors into account:
- If the port auto-negotiation is enabled,
  the effective speed should be the maximum speed in the configured advertised speeds list or the supported speeds list
- Otherwise, the effective speed is the configured speed.

#### How did you verify/test it?
Run regression test.

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

liat-grozovik commented on tests/platform/test_sfp.py:
```diff
@@ -13,7 +13,11 @@
+from loganalyzer import LogAnalyzer
 
+pytestmark = [
+    pytest.mark.disable_loganalyzer  # disable automatic loganalyzer
```
why disabling log analyzer? it should be enabled

liat-grozovik commented on tests/platform/test_sfp.py:
```diff
@@ -56,9 +60,14 @@ def test_check_sfp_status_and_configure_sfp(testbed_devices, conn_graph_facts):
+    loganalyzer = LogAnalyzer(ansible_host=ans_host, marker_prefix='sfp_cfg')
+    loganalyzer.load_common_config()
+
+    loganalyzer.ignore_regex.append("kernel.*Eeprom query failed*")
```
why are we ignoring errors from the eeprom instead of understanding why there are eeprom read failure? we should not have such errors in any case

liat-grozovik commented:
Please update the commit to refer to the bug and specific this is a WA till issue is fixed.

stephenxs commented:
community pr [https://github.com/Azure/sonic-mgmt/pull/1205] opened

stephenxs committed:
Fix the logic to fetch effective speed in headroom size calculation

Fix the logic to fetch the effective speed when calculating headroom size in the test case.
Originally, the effective speed is the configured speed.
However, with port auto negotiation supported, we need take more factors into account:
If the port auto negotiation is enabled, the effective speed should be the maximum speed
in the configured advertised speeds list or the supported speeds list
Otherwise, the effective speed is the configured speed.

Signed-off-by: Stephen Sun <stephens@nvidia.com>
